### PR TITLE
KIALI-1526 Show services nodes representing egress traffic

### PR DESCRIPTION
**Pull request by israel-hdez**
This is adding an exception to the dead node appender. This will allow terminal service nodes to be shown in the graph if those nodes are representing traffic going outside of the cluster.

** Backwards incompatible? **
- [ ] Is your change introducing backwards incompatible changes? **Nope**

### CODE REVIEW

**Comment by rhqci** [comment]
:heavy_check_mark: Jenkins CI: kiali-core-pr-e2e-test [#56](https://paste.fedoraproject.org/paste/og~LVcj79Z0fnXqu141Pqg)
  * :heavy_check_mark: Istio-and-Kiali-on-Openshift [#[1323]](https://paste.fedoraproject.org/paste/rOcCkApK5YA7O3XTYhZ5bA)
  * :heavy_check_mark: Istio-Upstream [#[205]](https://paste.fedoraproject.org/paste/4mx00tcCTfHEDm-N9~M68g)
  * :heavy_check_mark: Kiali-Istio-Checker [#[1174, 1175]](https://paste.fedoraproject.org/paste/ocvPD5nrcpRorbMDvhzshQ)
  * :heavy_check_mark: install-bookinfo [#[675]](https://paste.fedoraproject.org/paste/FrbOuwND~dBechBkBFxkUw)
  * :heavy_check_mark: Kiali-on-Openshift [#[693]](https://paste.fedoraproject.org/paste/eVOay09if-I99O-DLIuVLg)
  * :heavy_check_mark: run-kiali-e2e-tests [#[136]](https://paste.fedoraproject.org/paste/S~rMSLiWoH6mjy3Mmyz4zQ)


KIALI_NAME=kialicorepr506-istio-system

File: graph/appender/dead_node.go [hunk 1]
```diff
@@ -3,14 +3,20 @@ package appender
...
+	"github.com/kiali/kiali/services/models"
 )
 
 // DeadNodeAppender is responsible for removing from the graph unwanted nodes:
 // - nodes for which there is no traffic reported and the related schema is missing
 //   (presumably removed from K8S). (kiali-621)
 // - service nodes for which there is no incoming error traffic and no outgoing
-//   edges (kiali-1326)
-type DeadNodeAppender struct{}
+//   edges (kiali-1326) - excluding in this rule the nodes representing egress
```
Comment (jmazzitelli) [inline graph/appender/dead_node.go]: is this related to kiali-1326? I see "- excluding in this rule..." appears right after that. Do you mean to put "- excluding in this rule..." on a new line? Otherwise, its confusing - looks like this is related to kiali-1326 (unless this is related to 1326, in which case, ignore me :)

File: graph/appender/dead_node.go [hunk 2]
```diff
@@ -3,14 +3,20 @@ package appender
...
+//   traffic (KIALI-1526)
+type DeadNodeAppender struct {
+	// serviceEntries acts as a cache. This prevents from querying multiple times
+	// the cluser for service entries when the graph is requested for all namespaces.
```
Comment (jmazzitelli) [inline graph/appender/dead_node.go]: trivial - typo "cluster"

File: graph/appender/dead_node.go [hunk 3]
```diff
@@ -21,10 +27,52 @@ func (a DeadNodeAppender) AppendGraph(trafficMap graph.TrafficMap, _ string) {
...
+
+		serviceEntries = append(serviceEntries, istioCfg.ServiceEntries...)
+	}
+
+	return serviceEntries
+}
+
+// isExternalService checks whether node represents an external service (egress traffic), looking
+// for a matching egress rule in serviceEntries.
+func isExternalService(node *graph.Node, serviceEntries models.ServiceEntries) bool {
```
Comment (jmazzitelli) [inline graph/appender/dead_node.go]: Should this be passing by reference? (e.g. use "serviceEntries *models.ServiceEntries") --- this would avoid having to make a copy these entries every time we check a node. I'm not sure now large these ServiceEntries maps are, maybe its not a big deal. But it seems if we are going to call this for every node in the graph, we should think about passing this cache of service entries by reference.

**Comment by rhqci** [comment]
:heavy_check_mark: Jenkins CI: kiali-core-pr-e2e-test [#64](https://paste.fedoraproject.org/paste/ddooR~ifDI1e0tttqenvTA)
  * :red_circle: Istio-and-Kiali-on-Openshift [#[1335]](https://paste.fedoraproject.org/paste/xnAWLoe-YZe8hiYyK1XowA)
  * :heavy_check_mark: Istio-Upstream [#[215]](https://paste.fedoraproject.org/paste/RzVagb02-Bx40OZM6z6okQ)
  * :heavy_check_mark: Kiali-Istio-Checker [#[1198, 1199]](https://paste.fedoraproject.org/paste/GEZIzZ7r9DxAyQXvLEb3rw)
  * :heavy_check_mark: install-bookinfo [#[689]](https://paste.fedoraproject.org/paste/MV0ETorleIfqXN9X5nKFPw)
  * :heavy_check_mark: Kiali-on-Openshift [#[706]](https://paste.fedoraproject.org/paste/-tAtUmEzSA1JFQSYa1tbSA)
  * :red_circle: run-kiali-e2e-tests [#[151]](https://paste.fedoraproject.org/paste/5Gmv5tmwTTe~SJJtKkf6Tg)


KIALI_NAME=kialicorepr506-istio-system

File: graph/appender/dead_node.go (hunk 3 above)
Comment (israel-hdez) [reply graph/appender/dead_node.go]: Makes sense! Asterisk added.
Probably ServiceEntries may not be large... But who knows?... Only users I guess.

File: graph/appender/dead_node.go (hunk 1 above)
Comment (israel-hdez) [reply graph/appender/dead_node.go]: I think is related. In kiali-1326 _terminal service nodes_ that has NO errors are marked as dead nodes and excluded from the graph.

But egress traffic is recognized as service nodes in our code and they are terminal nodes. They matches with the logic of kiali-1326. So, in kiali-1526 I'm preventing these "egress nodes" from being identified as dead nodes.

File: graph/appender/dead_node.go (hunk 3 above)
Comment (jotak) [reply graph/appender/dead_node.go]: Actually slices already hold a pointer to the underlying array, so I think it wasn't necessary.

**Comment by israel-hdez** [comment]
Travis has hit twice that "sporadic test failure" in this PR: https://travis-ci.org/kiali/kiali/builds/431064612#L515

Do you want to check the logs? I though there was some logs written to the console. But I see nothing.

**Comment by jmazzitelli** [comment]
> Do you want to check the logs? I though there was some logs written to the console. But I see nothing.

@jshaughn added some new logs in the test, but I don't see it - i wonder if we need verbose logging enabled during the test runs?

File: graph/appender/dead_node.go (hunk 3 above)
Comment (israel-hdez) [reply graph/appender/dead_node.go]: @jotak I though that was the case. But added the reference just in case... It doesn't hurt, so ... I'll keep the asterisk :)

**Comment by rhqci** [comment]
:heavy_check_mark: Jenkins CI: kiali-core-pr-e2e-test [#67](https://paste.fedoraproject.org/paste/RkzFVeHI8NQ30D79B99j9A)
  * :red_circle: Istio-and-Kiali-on-Openshift [#[1339]](https://paste.fedoraproject.org/paste/~~xJp2qjeDJPzpf7~0NE0w)
  * :heavy_check_mark: Istio-Upstream [#[219]](https://paste.fedoraproject.org/paste/VVF0gH1ZbFK-Z5M5PdDulQ)
  * :heavy_check_mark: Kiali-Istio-Checker [#[1204, 1205]](https://paste.fedoraproject.org/paste/cezfc-9LVdDS85D4sgeuow)
  * :heavy_check_mark: install-bookinfo [#[692]](https://paste.fedoraproject.org/paste/OjhpmVV1~ivRHMkNv~PJww)
  * :heavy_check_mark: Kiali-on-Openshift [#[709]](https://paste.fedoraproject.org/paste/HGAwyTVlQJH9MPJcv1n4JQ)
  * :red_circle: run-kiali-e2e-tests [#[154]](https://paste.fedoraproject.org/paste/qp3Hmzfh5W8eyGsbvv1RZw)


KIALI_NAME=kialicorepr506-istio-system

**Comment by rhqci** [comment]
:heavy_check_mark: Jenkins CI: kiali-core-pr-e2e-test [#68](https://paste.fedoraproject.org/paste/Tamkt9pjBhq20vAEvO3Qqw)
  * :red_circle: Istio-and-Kiali-on-Openshift [#[1340]](https://paste.fedoraproject.org/paste/W2ujUaE3j5Hu8qWRzvIcag)
  * :heavy_check_mark: Istio-Upstream [#[220]](https://paste.fedoraproject.org/paste/PXrf7XRFufClkZOY29dibQ)
  * :heavy_check_mark: Kiali-Istio-Checker [#[1206, 1207]](https://paste.fedoraproject.org/paste/VdeDZTRu-zvux5JdVggpSA)
  * :heavy_check_mark: install-bookinfo [#[693]](https://paste.fedoraproject.org/paste/H-fYV3XFwK5cLEsFqPdQuQ)
  * :heavy_check_mark: Kiali-on-Openshift [#[710]](https://paste.fedoraproject.org/paste/lpcpSAY0UTuaQkk4UFob1w)
  * :red_circle: run-kiali-e2e-tests [#[155]](https://paste.fedoraproject.org/paste/9wqPqwOlHO4xj-f5lx6MiA)


KIALI_NAME=kialicorepr506-istio-system

**Comment by israel-hdez** [comment]
@jshaughn @jmazzitelli I see where is the problem in the sporadic failure.
I'll fix the problem in it's jira KIALI-1429. Please, merge if you see this PR ok.

File: graph/appender/dead_node.go [hunk 4]
```diff
@@ -37,7 +85,7 @@ func applyDeadNodes(trafficMap graph.TrafficMap, wkSvc business.WorkloadService)
 			rate4xx, hasRate4xx := n.Metadata["rate4xx"]
 			rate5xx, hasRate5xx := n.Metadata["rate5xx"]
 			numOutEdges := len(n.Edges)
-			if numOutEdges == 0 && (!hasRate4xx || rate4xx.(float64) == 0) && (!hasRate5xx || rate5xx.(float64) == 0) {
+			if numOutEdges == 0 && (!hasRate4xx || rate4xx.(float64) == 0) && (!hasRate5xx || rate5xx.(float64) == 0) && !isExternalService(n, serviceEntries) {
```
Comment (jshaughn) [inline graph/appender/dead_node.go]: If you chose to do the map of external mesh hosts I'd rather see that quick check done first.  if it's an external host just continue, and don't have the check appended to this long conditional.

File: graph/appender/dead_node.go [hunk 5]
```diff
@@ -3,14 +3,20 @@ package appender
...
 //   (presumably removed from K8S). (kiali-621)
 // - service nodes for which there is no incoming error traffic and no outgoing
-//   edges (kiali-1326)
-type DeadNodeAppender struct{}
+//   edges (kiali-1326) - excluding in this rule the nodes representing egress
+//   traffic (KIALI-1526)
+type DeadNodeAppender struct {
+	// serviceEntries acts as a cache. This prevents from querying multiple times
+	// the cluster for service entries when the graph is requested for all namespaces.
+	serviceEntries models.ServiceEntries
```
[the remark this answers is not in the document]
Comment (jotak) [reply graph/appender/dead_node.go]: > Go passes arrays by value

Yes, but this isn't an array, it's a slice, which is basically a pointer to an array; see https://blog.golang.org/go-slices-usage-and-internals . Using a pointer here would add a level of indirection. It doesn't hurt too much to do it when we're unsure but, I think, this is the kind of things we'll avoid in the future as long as language expertise go up :)

[the remark this answers is not in the document]
Comment (jshaughn) [reply graph/appender/dead_node.go]: Good point @jotak.  Slices are also passed by value but the size is small, it's just a pointer to the array and a little bit of header information, so pass by value is fine.  So ignore my comment, I still get confused sometimes with Go's syntax, I see [] and I think array.   But, I still recommend actually changing the whole thing to a map for quick lookups.  No need to do a looping search each time you want to know if a node is an external service.

**Comment by israel-hdez** [comment]
Ok, I changed to a map for performance

File: graph/appender/dead_node.go [hunk 6]
```diff
@@ -21,16 +27,54 @@ func (a DeadNodeAppender) AppendGraph(trafficMap graph.TrafficMap, _ string) {
...
+
+	applyDeadNodes(trafficMap, business.Workload, a.externalServices)
 }
 
-func applyDeadNodes(trafficMap graph.TrafficMap, wkSvc business.WorkloadService) {
+// resolveExternalServices queries the cluster API to resolve egress services
+// by using ServiceEntries resources across all namespaces in the cluster.
+// All ServiceEntries are needed, because Istio does not distinguish where the
+// ServiceEntries are created when routing egress traffic.
+func resolveExternalServices(bLayer *business.Layer) map[string]bool {
```
Comment (jshaughn) [inline graph/appender/dead_node.go]: We maybe need a way to cache this sort of information to be used repeatedly, and not redo it each time we generate a graph.  This is not a blocker, I'm just saying we need to maybe think about how to improve the handling for this slow-changing data.

File: graph/cytoscape/cytoscape.go [hunk 7]
```diff
@@ -52,6 +52,7 @@ type NodeData struct {
 	IsOutside       bool            `json:"isOutside,omitempty"`       // true | false
 	IsRoot          bool            `json:"isRoot,omitempty"`          // true | false
 	IsUnused        bool            `json:"isUnused,omitempty"`        // true | false
+	IsEgress        bool            `json:"isEgress,omitempty"`        // true | false
```
Comment (jshaughn) [inline graph/cytoscape/cytoscape.go]: Please maintain the alphabetical ordering.

**Comment by rhqci** [comment]
:heavy_check_mark: Jenkins CI: kiali-core-pr-e2e-test [#69](https://paste.fedoraproject.org/paste/6KMJ7c8-UZn9XPc7A2goBg)
  * :red_circle: Istio-and-Kiali-on-Openshift [#[1343]](https://paste.fedoraproject.org/paste/-C4AzKuT6VzPTG5g7fq47w)
  * :heavy_check_mark: Istio-Upstream [#[223]](https://paste.fedoraproject.org/paste/5nXEe5h-H4a8MRuuYsv12g)
  * :heavy_check_mark: Kiali-Istio-Checker [#[1219, 1220]](https://paste.fedoraproject.org/paste/eCljt6dIZKGgFF0BHVtUsA)
  * :heavy_check_mark: install-bookinfo [#[696]](https://paste.fedoraproject.org/paste/98FXQe77iTcpof3IVbl2ng)
  * :heavy_check_mark: Kiali-on-Openshift [#[713]](https://paste.fedoraproject.org/paste/vfxzj~cVvZ70S3rgRNwD0A)
  * :red_circle: run-kiali-e2e-tests [#[158]](https://paste.fedoraproject.org/paste/G36R2m~hdAzD45JTduar0g)


KIALI_NAME=kialicorepr506-istio-system

**Comment by rhqci** [comment]
:heavy_check_mark: Jenkins CI: kiali-core-pr-e2e-test [#70](https://paste.fedoraproject.org/paste/icWlevLORG43-42XmcsD1A)
  * :red_circle: Istio-and-Kiali-on-Openshift [#[1345]](https://paste.fedoraproject.org/paste/2s41J1GYyzLS5MDg88pxFw)
  * :heavy_check_mark: Istio-Upstream [#[225]](https://paste.fedoraproject.org/paste/MdC09lJz7DZgeknJUCLDjw)
  * :heavy_check_mark: Kiali-Istio-Checker [#[1224, 1225]](https://paste.fedoraproject.org/paste/lsivVZ0n2hFpJ~DHf9t~Sg)
  * :heavy_check_mark: install-bookinfo [#[699]](https://paste.fedoraproject.org/paste/4gGFcVXat76Z6~e8R~X8BQ)
  * :heavy_check_mark: Kiali-on-Openshift [#[715]](https://paste.fedoraproject.org/paste/CZs3iEluckC5q6EJDOuIkQ)
  * :red_circle: run-kiali-e2e-tests [#[161]](https://paste.fedoraproject.org/paste/qdMqtHol8-yJDcj-BDQdMw)


KIALI_NAME=kialicorepr506-istio-system

**Comment by rhqci** [comment]
:heavy_check_mark: Jenkins CI: kiali-core-pr-e2e-test [#71](https://paste.fedoraproject.org/paste/XvQ8QFKZsX-3OKBQaXV85Q)
  * :red_circle: Istio-and-Kiali-on-Openshift [#[1347]](https://paste.fedoraproject.org/paste/KuRNJx30dZ-Yk8IktEQuKg)
  * :heavy_check_mark: Istio-Upstream [#[227]](https://paste.fedoraproject.org/paste/Nv00liOqvvGHbuDeN8Q2GA)
  * :heavy_check_mark: Kiali-Istio-Checker [#[1228, 1229]](https://paste.fedoraproject.org/paste/lAd1yvuj7BjpLYhmHOfoyQ)
  * :heavy_check_mark: install-bookinfo [#[701]](https://paste.fedoraproject.org/paste/uv3tot9iifpXnKacXAFCMA)
  * :heavy_check_mark: Kiali-on-Openshift [#[717]](https://paste.fedoraproject.org/paste/~ujBdXCwUYPTAxyzOVQrcQ)
  * :red_circle: run-kiali-e2e-tests [#[163]](https://paste.fedoraproject.org/paste/Qyb~2pgL1gMBbpFm~NCDBA)


KIALI_NAME=kialicorepr506-istio-system

**Comment by rhqci** [comment]
:heavy_check_mark: Jenkins CI: kiali-core-pr-e2e-test [#73](https://paste.fedoraproject.org/paste/KjIMo7WyviPnPQfYpD6BAw)
  * :red_circle: Istio-and-Kiali-on-Openshift [#[1349]](https://paste.fedoraproject.org/paste/ypmHFuDy6l3LdoKof~oo9A)
  * :heavy_check_mark: Istio-Upstream [#[229]](https://paste.fedoraproject.org/paste/hKf8xBjz2R~C23hKm9QkDA)
  * :heavy_check_mark: Kiali-Istio-Checker [#[1232, 1233]](https://paste.fedoraproject.org/paste/RFQalCdrsgcYRpxoGDCKng)
  * :heavy_check_mark: install-bookinfo [#[703]](https://paste.fedoraproject.org/paste/~8QAJywHoF0S0qIcVxmF8Q)
  * :heavy_check_mark: Kiali-on-Openshift [#[719]](https://paste.fedoraproject.org/paste/zjqXCQkJXQAByZHlG5WTfg)
  * :red_circle: run-kiali-e2e-tests [#[165]](https://paste.fedoraproject.org/paste/k4ApUR5KwXnILlEvtUPUCA)


KIALI_NAME=kialicorepr506-istio-system
